e2e privacy policy

latin-panda opened the pull request:
# Description

Migration of privacy policy e2e test

https://github.com/medic/angular10-migration/issues/67

# Code review checklist
<!-- Remove or comment out any items that do not apply to this PR; in the remaining boxes, replace the [ ] with [x]. -->
- [ ] Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- [ ] Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- [ ] Tested: Unit and/or e2e where appropriate
- [ ] Internationalised: All user facing text
- [ ] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

latin-panda commented:
Hi @newtewt can you please have a look and let me know any feedback? Thank you!

newtewt commented on tests/e2e/login/privacy-policy.spec.js:
```diff
@@ -76,109 +74,122 @@ describe('Privacy policy', () => {
+      // After first login, check that privacy policy was prompted to user
+      await commonElements.goToLoginPageNative();
+      await loginPage.loginNative('online', password);
+      await helper.waitElementToPresentNative(element(by.css('#privacy-policy-wrapper')));
```
Can you move the elements to a page object? I think they would work in the `login.po.js`

latin-panda replied:
After taking look, I decided to move this into a new page object, privacy-policy.po since privacy policy was developed as an independent page and not closely related to login.

newtewt commented on tests/e2e/login/privacy-policy.spec.js:
```diff
@@ -76,109 +74,122 @@ describe('Privacy policy', () => {
+      // After first login in french, check that privacy policy was prompted to user
+      await commonElements.goToLoginPageNative();
+      await loginPage.loginNative('offline', password, false, 'fr');
+      await browser.sleep(1000); // overriding default language.
```
I think the call to brower.sleep is a bit flakey and the next call to waitElementToPresent might be a bit off since it only checks the element is in the dom. Might use `waitUntilReadyNative` as it checks for the element to be visible before continuing.

latin-panda replied:
Removed!

newtewt commented on tests/helper.js:
```diff
@@ -117,6 +138,26 @@ module.exports = {
+        12000,
+        `Element taking too long to appear in the DOM. Giving up! ${element.locator()}`
+      );
+      return await element.getText();
```
The await here should be able to be removed. The callers to this function are handling the promise by awaiting.

latin-panda replied:
Fixed!

newtewt reviewed (changes requested):
nice! Just a few minor changes.

newtewt commented on tests/helper.js:
```diff
@@ -89,7 +90,27 @@ module.exports = {
+
+    await elements.each(async (element) => {
+      await element.getText();
+      const text  = text.toLowerCase().trim();
```
text is being called before being set to anything here.

latin-panda replied:
Haha ups, fixed!

latin-panda commented:
Hi @newtewt, I've addressed the feedback and updated the branch with latest from take-2. Can you please take a look again? Thanks!

newtewt commented on tests/e2e/login/privacy-policy.spec.js:
```diff
@@ -76,109 +75,95 @@ describe('Privacy policy', () => {
-    helper.waitElementToBeVisible(element(by.css('.modal-dialog .modal-footer .btn')));
-    element(by.css('.modal-dialog .modal-footer .btn')).click();
+  const closeTourModal = async () => {
+    await helper.waitElementToBeVisibleNative(element(by.css('.modal-dialog .modal-footer .btn')));
```
Sorry I missed this one. We have a `utils.closeTour()` can that function replace this?

latin-panda replied:
Sure, works fine

newtewt commented:
Nice creating the new privacy policy page object. Cleans up the test nicely. I missed one thing during the first about reusing an existing function. Can you update?

latin-panda commented:
Hi @newtewt thank you for the review, I've added `the utils.closeTour() `
Can you please take a look again? 💪

newtewt commented:
coooooool! Looks good.

dianabarsan commented on tests/utils.js:
```diff
@@ -507,6 +509,26 @@ const createUsers = async (users, meta = false) => {
   }
 };
 
+const createUsersNative = async (users, meta = false) => {
```
I've just merged a PR that removes the need for such methods in this utils file. You can just use the existent methods and they should work fine.

latin-panda replied:
Done

dianabarsan commented on tests/utils.js:
```diff
@@ -899,6 +921,7 @@ module.exports = {
   revertDbNative,
 
   resetBrowser: () => {
+    deprecated('utils.resetBrowser', 'utils.resetBrowserNative');
```
This function is not deprecated anymore.

latin-panda replied:
Done

dianabarsan reviewed (changes requested):
Just a small request to use existent function, where it's possible.

latin-panda commented:
@dianabarsan I've addressed your feedback, but noticed that we still use the 'native' approach in helper. So I can removed all the natives from helper here or in a next PR to fix all the occurrences. I think in a next PR would be nice so it's all about removing the native while this one is about making privacy policy enable. But you guys let me know :)

dianabarsan commented:
> So I can removed all the natives from helper here or in a next PR to fix all the occurrences

We'll do that once all the tests have been migrated off of protractor's control flow, like @newtewt planned.

dianabarsan reviewed (approved):
Nice work!
